### PR TITLE
GUACAMOLE-903: Improved Chinese internationalization support

### PR DESCRIPTION
**Pull request by XYUU**
https://github.com/apache/guacamole-client/pull/453
https://github.com/apache/guacamole-client/pull/518

### CODE REVIEW

**Comment by necouchman** [comment]
@XYUU : You do not need to close and open new pull requests after each change that you make.  Simply pushing commits - either rebase and force-push or push new commits - will update the pull request automatically.  When you close existing requests and open new ones we lose the ability to quickly refer to previous conversations and requested changes, so, overall it is preferable to keep a single Pull Request open while we work through the issues toward merging it.

Thank you, again, for the contributions!

**Comment by mike-jumper** [comment]
Glad to see one commit this time around, but the merge base has changed back to `master`. We can change this to `staging/1.2.0` on our end, but once that's done things will likely need to be rebased due to commits from `master` appearing.

**Comment by mike-jumper** [comment]
We've bumped GUACAMOLE-903 from 1.2.0 so that the release can move forward with a first RC, so rebasing, etc. is no longer an issue and I have changed the merge base back to `master`.

**Comment by XYUU** [comment]
> The `guac-manifest.json` files of the CAS, Duo, LDAP, OpenID, Quick Connect, RADIUS, and TOTP extensions have been updated to point to `zh.json` files which do not exist.
> 
> Were these files intended to be committed with these changes?

I have committed

**Comment by necouchman** [comment]
@XYUU: Unfortunately this is still not quite right.  You still have 4 different commits, three of which have the exact same commit message ("Improved Chinese Internationalization Support") and one of which is a merge of the master back into your branch.  So, you still need to do the following:
- Rebase all of your changes on top of the `master` branch from the Apache repo, such that the only commits in the pull request are your changes.
- Either collapse all of your commits into a single commit, or update the commit messages such that they describe the actual changes being made in each of the commits instead of a generic message in all three of them.

**Comment by XYUU** [comment]
Well, I did it according to the requirements. Before that, I didn’t have the habit of force pressing. This is the first time I used force pushing.

**Comment by mike-jumper** [comment]
Hi @XYUU, the commits for this PR still show two duplicate messages:

Duplicate commits

As @necouchman mentioned, can you please:

> Either collapse all of your commits into a single commit, or update the commit messages such that they describe the actual changes being made in each of the commits instead of a generic message ...

Thanks!

Reviewing the translation changes themselves, things look fine to me. The only issue I see is with the existence of those two duplicate commits.

**Comment by XYUU** [comment]
> Hi @XYUU, the commits for this PR still show two duplicate messages:
> 
> Duplicate commits
> 
> As @necouchman mentioned, can you please:
> 
> > Either collapse all of your commits into a single commit, or update the commit messages such that they describe the actual changes being made in each of the commits instead of a generic message ...
> 
> Thanks!
> 
> Reviewing the translation changes themselves, things look fine to me. The only issue I see is with the existence of those two duplicate commits.

I also don't want to have two identical submissions. This is because after 4 months, when I checked whether the request was merged, I found a code conflict, and then I resolved the conflict. Since these two submissions are not adjacent, I cannot merge them into one submission.

Is there a better way? I want to hear your suggestions.

**Comment by mike-jumper** [comment]
You can definitely combine them. The easiest way in my opinion is to use git's interactive rebase feature and use the "fixup" operation to meld one commit into another.

You could also simply do a `git reset --soft HEAD~1` to effectively roll things back to just before the latest commit (`HEAD` is the latest commit and `HEAD~1` is that commit's parent) and then `git commit --amend` to add those changes to the most recent commit.

If you're worried about accidentally wiping out your changes:

* You can always make a backup branch to ensure you can go back to where you were.
* You can always reset things to the state of the pull request (the state of `master` in your fork)
* You can verify that things match before doing a force push by doing a `git diff` against the remote. The changes should be identical, even though there is one commit instead of two.

**Comment by XYUU** [comment]
> You can definitely combine them. The easiest way in my opinion is to use git's interactive rebase feature and use the "fixup" operation to meld one commit into another.
> 
> You could also simply do a `git reset --soft HEAD~1` to effectively roll things back to just before the latest commit (`HEAD` is the latest commit and `HEAD~1` is that commit's parent) and then `git commit --amend` to add those changes to the most recent commit.
> 
> If you're worried about accidentally wiping out your changes:
> 
> * You can always make a backup branch to ensure you can go back to where you were.
> * You can always reset things to the state of the pull request (the state of `master` in your fork)
> * You can verify that things match before doing a force push by doing a `git diff` against the remote. The changes should be identical, even though there is one commit instead of two.
image

Do you think this is appropriate?

**Comment by XYUU** [comment]
Thanks, you are great, I learned new knowledge!

File: extensions/guacamole-auth-cas/src/main/resources/guac-manifest.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
 
-    "guacamoleVersion" : "1.2.0",
+    "guacamoleVersion" : "1.3.0",
```
Comment (necouchman) [inline extensions/guacamole-auth-cas/src/main/resources/guac-manifest.json:3]: I'm not sure this should be done as part of this change. It definitely needs to be done for the overall 1.3.0 release, but it seems out-of-scope, here.

**Review by necouchman — CHANGES_REQUESTED** [review]

File: guacamole/src/main/webapp/translations/zh.json [hunk 2]
```diff
@@ -3,7 +3,6 @@
     "NAME" : "简体中文",
     
     "APP" : {
-
```
Comment (necouchman) [inline guacamole/src/main/webapp/translations/zh.json:6]: I think this blank line should remain there and not be removed - the rest of the sections (below) and other translations have these blank lines for readability.

**Review by necouchman — CHANGES_REQUESTED** [review]
I have one minor nitpick, but looks pretty good to me.

@mike-jumper any other concerns?

**Review by mike-jumper — APPROVED** [review]

**Comment by mike-jumper** [comment]
> I have one minor nitpick, but looks pretty good to me.
> 
> @mike-jumper any other concerns?

Nope - looks good to me. Once you're happy with your most recent feedback, I think we're good to go here.

File: guacamole/src/main/webapp/translations/zh.json (hunk 2 above)
Comment (necouchman) [reply guacamole/src/main/webapp/translations/zh.json:6]: @XYUU - If you can add back this poor space, I think we can move forward with the merge.

Comment (XYUU) [reply guacamole/src/main/webapp/translations/zh.json:6]: Okay, I'll take the time to add these spaces.

File: guacamole/src/main/webapp/translations/zh.json [hunk 3]
```diff
@@ -300,14 +319,129 @@
...
         "SECTION_HEADER_EDIT_USER"   : "编辑用户",
         "SECTION_HEADER_PERMISSIONS" : "使用权限",
-
+        "SECTION_HEADER_USER_GROUPS"         : "用户组",
+	
         "TEXT_CONFIRM_DELETE" : "将不能恢复已被删除的用户。确定要删除这个用户吗？"
 
+    },
+    
+        "MANAGE_USER_GROUP" : {
```
Comment (necouchman) [inline guacamole/src/main/webapp/translations/zh.json:338]: This has been indented one level too much...

File: guacamole/src/main/webapp/translations/zh.json [hunk 4]
```diff
@@ -490,31 +649,40 @@
 
     },
 
+
+
```
Comment (necouchman) [inline guacamole/src/main/webapp/translations/zh.json:653]: I don't think we need the extra spaces, here - the original one was fine :-).

**Review by necouchman — CHANGES_REQUESTED** [review]
A couple of more style issues...almost there...

**Comment by mike-jumper** [comment]
By the way:

* @XYUU, thanks for your patience as we work with you to mold things to match established style.
* @necouchman, I think we should probably look toward some automated means of re-indenting, sorting, grouping, etc. the translation JSON to reduce the back-and-forth required for contributed translations. That, or perhaps leverage some external service for contributing translations which does not require technical expertise and experience with git. Lowering the barrier for contributing translations would no doubt lead to more supported languages and more consistently up-to-date text.

**Comment by necouchman** [comment]
@XYUU I echo @mike-jumper's comments, thank you for working with us through this.

@mike-jumper Agreed, this shouldn't need to be a manual process.

**Comment by necouchman** [comment]
@XYUU Any chance you can finish up these minor changes so we can get this merged?

**Comment by XYUU** [comment]
> @XYUU Any chance you can finish up these minor changes so we can get this merged?
Okay！

**Review by necouchman — APPROVED** [review]
@XYUU Thanks again for your patience in working through this. :shipit:

**Pull request merged**